### ISSUE TITLE
Developer experience and rough workflow

### ISSUE THREAD

**Issue by TrevorFSmith** [mozilla/action-input#1]
I've been thinking about the general shape and major classes of the libraries so that we can being to feel out what work is on the developers and what work the libraries take on. It seems like the devs will need to do some up-front work declaring action maps and creating any custom filters or input sources. Hopefully, we can provide default action maps, filters, and input sources so that the dev just extends them with a little custom work.

Here's a rough list of work the dev does during development and then work the code does at runtime. 

Please chime in if anything is missing, wrong, or needs tweaking.
  
## During development

- extend types JSON (if any)
- extend action maps JSON
- code custom Filters (if any)
- code custom InputSources (if any)

## At runtime

- instantiate ActionMaps from JSON
- instantiate Types from JSON
- instantiate Filters using the Types
- instantiate InputSources using the Types
- instantiate an ActionMapper with the ActionMaps, Filters, and InputSources
- instantiate an ActionManager with an ActionMapper
- subscribe to action events from the ActionManager
- poll the ActionManager during the game loop
- toggle ActionMaps based on:
	device changes: landscape, portrait, online, offline, ...
	mode changes: flat, portal, immersive
	situation changes: paused, playing, configuring

![xr input library data flow 002](https://user-images.githubusercontent.com/34779/36114238-46e2abde-0fe4-11e8-95e6-7c3a1a421b71.png)

**Comment by johnshaughnessy** [comment]
>Most of the people who show up won't be technical and they won't go looking for bindings that the nytimes didn't create.
The idea is that there will be a small number of people who will create, share, and upvote bindings so that there is a reasonable default available for everyone else. If even one binding exists for a user's device for an application, that can be used as the default. 

>are the configs per-domain? Per-page? Per unique URL, including parameters? Linked by an id that the web app developer has puts in a meta tag?
Good question... If we are hosting and sharing filters, perhaps the ID-in-the-meta-tag approach is best.

**Comment by fernandojsg** [comment]
I'm looking around the proposal to try to match the lower level implementation and I've some questions for @netpro2k and @johnshaughnessy:

In my prototype I was handling the buttons state changes at the lower level itself, and generating events when button down, up or change occurs. The main idea behind it is that, as we discussed initially, we could use that lower level API by itself without any of the extra action/mapping functionality on top of that, basically as a nicer abstraction over the current input apis (as shown here https://twitter.com/fernandojsg/status/963856765812117504)
But in your code the change status and even the query of the state of the input is done directly on the input source which is already in a middle/high level of abstraction and there we already will need semantic paths and so on. Is it that on purpose? If so, should we focus this implementation on the high level goal and forget about reusing the lower level as an stand alone library? Or should we try to match these both keeping the standalone low level adapted to your proposal?
If the later, one option could be to fill the inputFrame directly with the input data from the lower level, but we'll have two ways where we'll generate the "pressdown" event, on your code and on the lower level bindings. Should we use the lower level binding for that purpose instead?

I'll try to come out with an example of both proposals in the meantime but I would like to hear your thoughts on that

**Comment by fernandojsg** [comment]
TL;DR: should we focus on how we would like to use the higher level API and implement the lower level just to match that use, or should we think also on using the lower level as standalone and use it on the hightlevel as a third party component?

**Comment by TrevorFSmith** [comment]
My general plan was to have two separate wadges of functionality: one that maps low level hardware events into semantic events (e.g. button click to "jump" or recognized hand gesture to "open menu") and one that most apps talk to that provides an event API and a polling API for actions.

I think the next thing I'm going to do is to stub out the APIs of both wadges and submit a PR for discussion.
